compiler/passes/src/destructuring/destructure_expression.rs

```diff
@@ -37,28 +43,105 @@ impl ExpressionReconstructor for Destructurer<'_> {
             panic!("SSA guarantees that subexpressions are identifiers or literals.");
         };
 
-        // Lookup the expression in the tuple map.
-        match self.tuples.get(&identifier.name).and_then(|tuple_expr| tuple_expr.elements.get(input.index.value())) {
-            Some(element) => (element.clone(), Default::default()),
+        // Look up the expression in the tuple map.
+        match self.tuples.get(&identifier.name).and_then(|tuple_names| tuple_names.get(input.index.value())) {
+            Some(id) => (Expression::Identifier(*id), Default::default()),
             None => {
-                if matches!(self.type_table.get(&identifier.id), Some(Type::Future(_))) {
-                    (
-                        Expression::Access(AccessExpression::Array(ArrayAccess {
-                            array: Box::new(Expression::Identifier(*identifier)),
-                            index: Box::new(Expression::Literal(Literal::Integer(
-                                IntegerType::U32,
-                                input.index.to_string(),
-                                input.span,
-                                Default::default(),
-                            ))),
-                            span: input.span,
-                            id: input.id,
-                        })),
+                if !matches!(self.type_table.get(&identifier.id), Some(Type::Future(_))) {
+                    panic!("Type checking guarantees that all tuple accesses are declared and indices are valid.");
+                }
+
+                let expr = Expression::Access(AccessExpression::Array(ArrayAccess {
+                    array: Box::new(Expression::Identifier(*identifier)),
+                    index: Box::new(Expression::Literal(Literal::Integer(
+                        IntegerType::U32,
+                        input.index.to_string(),
+                        input.span,
                         Default::default(),
-                    )
+                    ))),
+                    span: input.span,
+                    id: input.id,
+                }));
+
+                (expr, Default::default())
+            }
+        }
+    }
+
+    fn reconstruct_ternary(&mut self, mut input: TernaryExpression) -> (Expression, Self::AdditionalOutput) {
```

Comment:
Could we add a comment to the top of the method describing the transformation? Some inline comments would be helpful as well.